common/network-common/src/main/java/org/apache/spark/network/sasl/aes/AesCipher.java

```diff
@@ -0,0 +1,340 @@
+    if (!inputStreamMap.containsKey(ch)) {
+      inputStreamMap.put(ch, new CryptoInputStream(TRANSFORM, properties, ch, inKeySpec, inIvSpec));
+    }
+
+    return inputStreamMap.get(ch);
+  }
+
+  /**
+   * Add handlers to channel
+   * @param ch the channel for adding handlers
+   * @throws IOException
+   */
+  public void addToChannel(Channel ch) throws IOException {
+    ch.pipeline()
+      .addFirst(ENCRYPTION_HANDLER_NAME, new AesEncryptHandler(this))
+      .addFirst(DECRYPTION_HANDLER_NAME, new AesDecryptHandler(this));
+  }
+
+  /**
+   * Generate a request config message which send to remote peer.
+   * @param conf is the local transport configuration.
+   * @return Config message for sending.
+   */
+  public static AesConfigMessage requestConfigMessage(TransportConf conf) {
+    int keySize = conf.saslEncryptionAesCipherKeySizeBits();
+    if (keySize % 8 != 0) {
+      throw new IllegalArgumentException("The AES cipher key size in bits should be a multiple " +
+        "of byte");
+    }
+    return new AesConfigMessage(keySize/8, null, null, null, null);
+  }
+
+  /**
+   * Generate the configuration message according to request config message.
+   * @param configMessage The request config message comes from remote.
+   * @return Configuration message for sending.
+   */
+  public static AesConfigMessage responseConfigMessage(AesConfigMessage configMessage){
+
+    Properties properties = new Properties();
+    int keyLen = configMessage.keySize;
+
+    try {
+      int paramLen = CryptoCipherFactory.getCryptoCipher(AesCipher.TRANSFORM, properties)
+        .getBlockSize();
+      byte[] inKey = new byte[keyLen];
+      byte[] outKey = new byte[keyLen];
+      byte[] inIv = new byte[paramLen];
+      byte[] outIv = new byte[paramLen];
+
+      CryptoRandom random = CryptoRandomFactory.getCryptoRandom(properties);
+      random.nextBytes(inKey);
+      random.nextBytes(outKey);
+      random.nextBytes(inIv);
+      random.nextBytes(outIv);
+
+      configMessage.setParameters(keyLen, inKey, inIv, outKey, outIv);
+    } catch (Exception e) {
+      logger.error("AES negotiation exception", e);
+      throw Throwables.propagate(e);
+    }
+
+    return configMessage;
+  }
+
+  private static class AesEncryptHandler extends ChannelOutboundHandlerAdapter {
+    private ByteArrayWritableChannel byteChannel;
+    private CryptoOutputStream cos;
+
+    AesEncryptHandler(AesCipher cipher) throws IOException {
+      byteChannel = new ByteArrayWritableChannel(AesCipher.STREAM_BUFFER_SIZE);
+      cos = cipher.CreateOutputStream(byteChannel);
+    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
+      throws Exception {
+      ctx.write(new EncryptMessage(cos, msg, byteChannel), promise);
+    }
+
+    @Override
+    public void close(ChannelHandlerContext ctx, ChannelPromise promise) throws Exception {
+      cos.close();
```

Comment:
Both of them will throw exception, if use "try...finally", exception will be throw in finally block.